Codegen checks concentrations probabilities

MauriceHendrix opened the pull request:
Made changes to tests for codegen (in heart ) to go ith the changes in codegen that noe check state variables for nan/info and check probabilities and concentrations are in expected ranges

I've put some review comments in to explain a bit more.
These pass Continuous, Nightly, Weekly (and Codegen) test packs locally.

MauriceHendrix committed:
updated codegen to include checks for nan/info and probabilities not between 0 and 1 and negative concentrations

MauriceHendrix committed:
fix Shannon 2004 tests

MauriceHendrix committed:
Merge branch 'develop' of https://github.com/Chaste/Chaste into codegen_checks_concentrations_probabilities

mirams commented:
Talk me through why the Shannon thing changed, don’t really want to increase number of timesteps by 10x, it’s a performance benchmark test!

MauriceHendrix commented:
> Talk me through why the Shannon thing changed, don’t really want to increase number of timesteps by 10x, it’s a performance benchmark test!

because if we don't it'll throw an error as a state variable gets out of range.

> Talk me through why the Shannon thing changed, don’t really want to increase number of timesteps by 10x, it’s a performance benchmark test!

UpdateCTestConfiguration  from :/home/uczmh2/develop/chaste_build/DartConfiguration.tcl
UpdateCTestConfiguration  from :/home/uczmh2/develop/chaste_build/DartConfiguration.tcl
Test project /home/uczmh2/develop/chaste_build
Constructing a list of tests
Done constructing a list of tests
Updating test list for fixtures
Added 0 tests to meet fixture requirements
Checking test dependency graph...
Checking test dependency graph end
test 400
    Start 400: Test1dMonodomainShannonCvodeBenchmarks

400: Test command: /home/uczmh2/develop/chaste_build/heart/test/Test1dMonodomainShannonCvodeBenchmarks
400: Test timeout computed to be: 9.99988e+06
400: Running 1 test
400: 
400:  ***** Test1dMonodomainShannonCvodeBenchmarks.hpp *****
400: 
400: 
400: Entering TestWithDifferentCellsAndSolvers
400: 1. Forward Euler for 25 ms took time: 3.43161s
400: Global node 4 had problems with ODE solve between t = 7 and 7.100000000000001ms.
400: Voltage at this node before solve was 31.26909090406994mV
400: (this SHOULD NOT necessarily be the same as the one in the state variables,
400: which can be ignored and stay at the initial condition - the voltage is dictated by PDE instead of state variable.)
400: Stimulus current (NB converted to micro-Amps per cm^3) applied here is equal to:
400: 	0 at t = 7ms,
400: 	0 at t = 7.100000000000001ms.
400: Cell model: shannon_2004
400: All state variables are now:
400: 	membrane_voltage	:	31.26909090406994
400: 	cytosolic_calcium_concentration	:	9.657962175160678e-05
400: 	membrane_fast_sodium_current_h_gate	:	-1.021282259066593e-11
400: 	membrane_fast_sodium_current_j_gate	:	0.3367547045719115
400: 	membrane_fast_sodium_current_m_gate	:	0.9998203520087352
400: 	IKr_Xr_gate__Xr	:	0.1991276016183594
400: 	IKs_Xs_gate__Xs	:	0.02015389202722888
400: 	Itos_X_gate__X_tos	:	0.8907283020757806
400: 	Itos_Y_gate__Y_tos	:	0.8864172447774213
400: 	Itos_R_gate__R_tos	:	0.978738330497301
400: 	Itof_X_gate__X_tof	:	0.7277238938187823
400: 	Itof_Y_gate__Y_tof	:	0.8322916951512394
400: 	membrane_L_type_calcium_current_d_gate	:	0.9975070949383857
400: 	membrane_L_type_calcium_current_f_gate	:	0.9387758937265587
400: 	ICaL_fCa_gate__fCaB_SL	:	0.01820576095927897
400: 	ICaL_fCa_gate__fCaB_jct	:	0.03784647157834065
400: 	Jrel_SR__R	:	0.8674386216267038
400: 	Jrel_SR__I	:	4.990430448915094e-06
400: 	Jrel_SR__O	:	3.266173455575902e-05
400: 	Na_buffer__Na_SL	:	8.814041351503709
400: 	Na_buffer__Na_jct	:	8.848434519783895
400: 	Na_buffer__Na_SL_buf	:	0.7721038935944624
400: 	Na_buffer__Na_jct_buf	:	3.540293112966086
400: 	cytosolic_sodium_concentration	:	8.815400925155968
400: 	JSR_calcium_concentration	:	0.5762008740207379
400: 	Ca_buffer__Ca_SL	:	0.0003112565439345414
400: 	Ca_buffer__Ca_jct	:	0.004281527579359794
400: 	Ca_buffer__Ca_SLB_SL	:	0.0279583273835082
400: 	Ca_buffer__Ca_SLB_jct	:	0.1152724041930639
400: 	Ca_buffer__Ca_SLHigh_SL	:	0.1451441604524701
400: 	Ca_buffer__Ca_SLHigh_jct	:	0.1202679223967679
400: 	Ca_buffer__Ca_Calsequestrin	:	1.221760767815086
400: 	cytosolic_Ca_buffer__Ca_TroponinC	:	0.008946156853538464
400: 	cytosolic_Ca_buffer__Ca_TroponinC_Ca_Mg	:	0.1078320627641118
400: 	cytosolic_Ca_buffer__Mg_TroponinC_Ca_Mg	:	0.01524029300317839
400: 	cytosolic_Ca_buffer__Ca_Calmodulin	:	0.0003237253446141173
400: 	cytosolic_Ca_buffer__Ca_Myosin	:	0.001299396893155021
400: 	cytosolic_Ca_buffer__Mg_Myosin	:	0.1381982986321819
400: 	cytosolic_Ca_buffer__Ca_SRB	:	0.002250387215810784
400: terminate called after throwing an instance of 'Exception'
400: [pmlapgmi01:21069] *** Process received signal ***
400: [pmlapgmi01:21069] Signal: Aborted (6)
400: [pmlapgmi01:21069] Signal code:  (-6)
400: [pmlapgmi01:21069] [ 0] /lib/x86_64-linux-gnu/libc.so.6(+0x3ef10)[0x7f8504d81f10]
400: [pmlapgmi01:21069] [ 1] /lib/x86_64-linux-gnu/libc.so.6(gsignal+0xc7)[0x7f8504d81e87]
400: [pmlapgmi01:21069] [ 2] /lib/x86_64-linux-gnu/libc.so.6(abort+0x141)[0x7f8504d837f1]
400: [pmlapgmi01:21069] [ 3] /usr/lib/x86_64-linux-gnu/libstdc++.so.6(+0x8c957)[0x7f850590a957]
400: [pmlapgmi01:21069] [ 4] /usr/lib/x86_64-linux-gnu/libstdc++.so.6(+0x92ae6)[0x7f8505910ae6]
400: [pmlapgmi01:21069] [ 5] /usr/lib/x86_64-linux-gnu/libstdc++.so.6(+0x92b21)[0x7f8505910b21]
400: [pmlapgmi01:21069] [ 6] /usr/lib/x86_64-linux-gnu/libstdc++.so.6(+0x92d54)[0x7f8505910d54]
400: [pmlapgmi01:21069] [ 7] /home/uczmh2/develop/chaste_build/heart/libchaste_heart.so(_ZN22AbstractCardiacProblemILj1ELj1ELj1EE5SolveEv+0x10a9)[0x7f8509ae51f5]
400: [pmlapgmi01:21069] [ 8] /home/uczmh2/develop/chaste_build/heart/test/Test1dMonodomainShannonCvodeBenchmarks(+0x1a816)[0x564c1c50a816]
400: [pmlapgmi01:21069] [ 9] /home/uczmh2/develop/chaste_build/heart/test/Test1dMonodomainShannonCvodeBenchmarks(+0x1b242)[0x564c1c50b242]
400: [pmlapgmi01:21069] [10] /home/uczmh2/develop/chaste_build/heart/test/Test1dMonodomainShannonCvodeBenchmarks(+0x13eab)[0x564c1c503eab]
400: [pmlapgmi01:21069] [11] /home/uczmh2/develop/chaste_build/heart/test/Test1dMonodomainShannonCvodeBenchmarks(+0x16983)[0x564c1c506983]
400: [pmlapgmi01:21069] [12] /home/uczmh2/develop/chaste_build/heart/test/Test1dMonodomainShannonCvodeBenchmarks(+0x16873)[0x564c1c506873]
400: [pmlapgmi01:21069] [13] /home/uczmh2/develop/chaste_build/heart/test/Test1dMonodomainShannonCvodeBenchmarks(+0x16723)[0x564c1c506723]
400: [pmlapgmi01:21069] [14] /home/uczmh2/develop/chaste_build/heart/test/Test1dMonodomainShannonCvodeBenchmarks(+0x165f1)[0x564c1c5065f1]
400: [pmlapgmi01:21069] [15] /home/uczmh2/develop/chaste_build/heart/test/Test1dMonodomainShannonCvodeBenchmarks(+0x16bb2)[0x564c1c506bb2]
400: [pmlapgmi01:21069] [16] /home/uczmh2/develop/chaste_build/heart/test/Test1dMonodomainShannonCvodeBenchmarks(+0x13005)[0x564c1c503005]
400: [pmlapgmi01:21069] [17] /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xe7)[0x7f8504d64c87]
400: [pmlapgmi01:21069] [18] /home/uczmh2/develop/chaste_build/heart/test/Test1dMonodomainShannonCvodeBenchmarks(+0x12eba)[0x564c1c502eba]
400: [pmlapgmi01:21069] *** End of error message ***
1/1 Test #400: Test1dMonodomainShannonCvodeBenchmarks ...***Exception: Child aborted  4.84 sec

0% tests passed, 1 tests failed out of 1

Label Time Summary:
Continuous_heart    =   4.84 sec*proc (1 test)

Total Test time (real) =   4.85 sec

The following tests FAILED:
	400 - Test1dMonodomainShannonCvodeBenchmarks (Child aborted)
Errors while running CTest

MauriceHendrix commented:
> Talk me through why the Shannon thing changed, don’t really want to increase number of timesteps by 10x, it’s a performance benchmark test!

Unfortunately the test fails if we don't because `membrane_fast_sodium_current_h_gate` becomes negative. Of course we can add yet another hack exception for thiss in codegen if really necessary.

```
400:  ***** Test1dMonodomainShannonCvodeBenchmarks.hpp *****
400: 
400: Entering TestWithDifferentCellsAndSolvers
400: 1. Forward Euler for 25 ms took time: 3.43161s
400: Global node 4 had problems with ODE solve between t = 7 and 7.100000000000001ms.
400: Voltage at this node before solve was 31.26909090406994mV
400: (this SHOULD NOT necessarily be the same as the one in the state variables,
400: which can be ignored and stay at the initial condition - the voltage is dictated by PDE instead of state variable.)
400: Stimulus current (NB converted to micro-Amps per cm^3) applied here is equal to:
400: 	0 at t = 7ms,
400: 	0 at t = 7.100000000000001ms.
400: Cell model: shannon_2004
400: All state variables are now:
400: 	membrane_voltage	:	31.26909090406994
400: 	cytosolic_calcium_concentration	:	9.657962175160678e-05
400: 	membrane_fast_sodium_current_h_gate	:	-1.021282259066593e-11
400: 	membrane_fast_sodium_current_j_gate	:	0.3367547045719115
400: 	membrane_fast_sodium_current_m_gate	:	0.9998203520087352
400: 	IKr_Xr_gate__Xr	:	0.1991276016183594
400: 	IKs_Xs_gate__Xs	:	0.02015389202722888
400: 	Itos_X_gate__X_tos	:	0.8907283020757806
400: 	Itos_Y_gate__Y_tos	:	0.8864172447774213
400: 	Itos_R_gate__R_tos	:	0.978738330497301
400: 	Itof_X_gate__X_tof	:	0.7277238938187823
400: 	Itof_Y_gate__Y_tof	:	0.8322916951512394
400: 	membrane_L_type_calcium_current_d_gate	:	0.9975070949383857
400: 	membrane_L_type_calcium_current_f_gate	:	0.9387758937265587
400: 	ICaL_fCa_gate__fCaB_SL	:	0.01820576095927897
400: 	ICaL_fCa_gate__fCaB_jct	:	0.03784647157834065
400: 	Jrel_SR__R	:	0.8674386216267038
400: 	Jrel_SR__I	:	4.990430448915094e-06
400: 	Jrel_SR__O	:	3.266173455575902e-05
400: 	Na_buffer__Na_SL	:	8.814041351503709
400: 	Na_buffer__Na_jct	:	8.848434519783895
400: 	Na_buffer__Na_SL_buf	:	0.7721038935944624
400: 	Na_buffer__Na_jct_buf	:	3.540293112966086
400: 	cytosolic_sodium_concentration	:	8.815400925155968
400: 	JSR_calcium_concentration	:	0.5762008740207379
400: 	Ca_buffer__Ca_SL	:	0.0003112565439345414
400: 	Ca_buffer__Ca_jct	:	0.004281527579359794
400: 	Ca_buffer__Ca_SLB_SL	:	0.0279583273835082
400: 	Ca_buffer__Ca_SLB_jct	:	0.1152724041930639
400: 	Ca_buffer__Ca_SLHigh_SL	:	0.1451441604524701
400: 	Ca_buffer__Ca_SLHigh_jct	:	0.1202679223967679
400: 	Ca_buffer__Ca_Calsequestrin	:	1.221760767815086
400: 	cytosolic_Ca_buffer__Ca_TroponinC	:	0.008946156853538464
400: 	cytosolic_Ca_buffer__Ca_TroponinC_Ca_Mg	:	0.1078320627641118
400: 	cytosolic_Ca_buffer__Mg_TroponinC_Ca_Mg	:	0.01524029300317839
400: 	cytosolic_Ca_buffer__Ca_Calmodulin	:	0.0003237253446141173
400: 	cytosolic_Ca_buffer__Ca_Myosin	:	0.001299396893155021
400: 	cytosolic_Ca_buffer__Mg_Myosin	:	0.1381982986321819
400: 	cytosolic_Ca_buffer__Ca_SRB	:	0.002250387215810784
400: terminate called after throwing an instance of 'Exception'
```

mirams commented:
Skimming that test, it appears that the normal output is something like 
```
Entering TestWithDifferentCellsAndSolvers
1. Forward Euler for 25 ms took time: 7.79142s
2. CVODE adaptor for 25 ms took time: 3.65384s
3. CVODE native for 25 ms took time: 2.15249s
Passed
```
And it seems to be passing the Forward Euler and failing in the second one if I read the output correctly.
 
So, the problem is probably us having a `VerifyStateVariables` with '<0' within the autogenerated 'normal' cells (which is sensible), but the cvode adaptor is being used (runs cvode solver on the 'normal' cells with the overhead of translating between vector types), so it should be checking with  '< -tol' rather than '<0' as the `VerifyStateVariables` methods for CVODE cells do!

CVODE Adaptor solver is something of an oddity for cardiac as `chaste_codegen` of course makes the specialised ODE system type with sundials' N_Vectors. So it was probably done here to prove it was worth generating and using special Cvode cells in tissue simulations. Something like a dynamic_cast on the `AbstractCardiacCell::mpOdeSolver` to check whether it is a `CvodeAdaptor`, and then if it is, don't call `VerifyStateVariables`, but would be a little bit annoying in terms of adding overhead for a `dynamic_cast` to all non-cvode cardiac solves, although only within the debug runs I suppose as that could be done within the `#ifndef NDEBUG` blocks within `AbstractCardiacCell` class.

mirams commented on heart/test/ionicmodels/TestCodegenLong/TestCodegenLongBackwardEuler.hpp:
```diff
@@ -72,7 +72,14 @@ class TestCodegenLongBackwardEuler : public CodegenLongHelperTestSuite
+        TS_ASSERT_THROWS_ANYTHING(RunTests(dirname, {"negative_concentration_paci_hyttinen_aaltosetala_severi_ventricularVersion"}, args));
+
+        // initial value for membrane_L_type_calcium_current_f_gate in Shannon2004 is slightly above 1 (it is a probability)
+        std::vector<std::string> bigger_tolerance_models = spectail_streatment_models(models, {"Shannon2004"});
```
Should this read `special_treatment_models` ?

mirams commented:
This isn't really a very nice way of dealing with the Shannon problem:
```
        SetUseCvOdeTolerances(1e-4);
        std::vector<std::string> bigger_tolerance_models = spectail_streatment_models(models, {"Shannon2004"});
        RunTests(dirname, bigger_tolerance_models, args);
```

as it is relying on solving the ODEs more coarsely putting a bigger tolerance in `VerifyStateVariables` and most users of Chaste aren't going to want to do that (including ApPredict which might start failing). 

Very strictly we should probably remove the metadata tag from the CellML for that variable in Shannon, because it isn't mathematically a probability any more (and we just updated the metadata to say anything with an f gate tag is a probability!). Only problem with that is it probably makes Shannon disappear from Aditi's stories and comparisons of gates etc. when it is quite an important calcium model. 

My proposal then, is to hack chaste_codegen where it generates the VerifyStateVariables to skip checking the upper bound of the calcium f gate variable if the model name is what's in the shannon cellml!

MauriceHendrix commented:
> 
> My proposal then, is to hack chaste_codegen where it generates the VerifyStateVariables to skip checking the upper bound of the calcium f gate variable if the model name is what's in the shannon cellml!

That's already in it, the tolerance stuff above isn't needed anymore.
But I don't like that one bit tbh as it's an awful hack. 

My prefered solution would be to introduce a non-probability gate (which is just a special gate) and use that to identify it. That wouldn't mess with Aditi's stuff or rely on hard-coding the model name.

mirams commented:
For posterity, the problem is that despite the Shannon L-type calcium current "f gate" looking like a Hodgkin-Huxley gating variable with the usual equation
<img width="396" alt="image" src="https://user-images.githubusercontent.com/858776/213762303-e7f2cd70-221f-4ddb-bbc3-bafa04f6fc1a.png">
the f_infinity is a bit non-standard
<img width="667" alt="image" src="https://user-images.githubusercontent.com/858776/213762352-5c6fedfa-562a-45d1-8c71-f92c920269df.png">
and actually goes above one, meaning the f gate can also go slightly above one!
<img width="1168" alt="image" src="https://user-images.githubusercontent.com/858776/213763000-4c4be450-f591-4736-9e98-204ca2ed9f2d.png">
so whilst it looks a lot like a Hodgkin-Huxley gate and therefore probability, it isn't quite...

mirams commented:
@MichaelClerx I think there were a couple of examples of non-probability gates called the same as gates in Aditi's survey, some even more marked than this I think. Can you remember which models, so we can see if that informs a sensible strategy?

MauriceHendrix committed:
typo and removing unneeded larger tolerance in test

MichaelClerx commented:
"Similarly, the models by Demir et al. (1994), Lindblad et al. (1996), and Aslanidi, Boyett, et al. (2009) allow O to be greater than one (so that the usual interpretation of open probability does not apply)"

Check the review for the full references :D
Especially for Aslanidi, as there are two Aslanidi 2009 models

MauriceHendrix commented:
Right so after checking it turns out Aditi's stories do use the "offending" gates in the Shannon2004 model. I have added exceptions to codegen.

mirams commented on heart/test/ionicmodels/TestCodegenLong/TestCodegenLongAnalyticCvode.hpp:
```diff
@@ -62,11 +62,19 @@ class TestCodegenLongAnalyticCvode : public CodegenLongHelperTestSuite
+        TS_ASSERT_THROWS_ANYTHING(RunTests(dirname, {"negative_concentration_paci_hyttinen_aaltosetala_severi_ventricularVersion"}, args));
+
+        // initial value for membrane_L_type_calcium_current_f_gate in Shannon2004 is slightly above 1 (it is a probability)
+        std::vector<std::string> bigger_tolerance_models = special_treatment_models(models, {"Shannon2004"});
```
We shouldn't need lines like this in the tests if codegen is no longer printing the `>1` check in the generated c++?

MauriceHendrix committed:
removed cvode precision

mirams commented on heart/test/ionicmodels/TestCodegenLong/TestCodegenLongBackwardEulerOpt.hpp:
```diff
@@ -61,17 +61,19 @@ class TestCodegenLongBackwardEulerOpt : public CodegenLongHelperTestSuite
+        std::vector<std::string> bad_models = special_treatment_models(models, {"hund_rudy_2004_a"});
 
         HeartConfig::Instance()->SetOdePdeAndPrintingTimeSteps(0.01, 0.1, 1.0);
+        TS_ASSERT_THROWS_ANYTHING(RunTests(dirname, {"negative_concentration_paci_hyttinen_aaltosetala_severi_ventricularVersion"}, args));
```
`TS_ASSERT_THROWS_CONTAINS` would be a bit nicer - checks that the exception is the one we are expecting

MauriceHendrix committed:
switched to cellml submodule and moved codegen concentration error check

MauriceHendrix committed:
moving to using cellml submodule

MauriceHendrix committed:
catching up with new cellml models in test

MauriceHendrix committed:
adjust heart tests to cellml files in cellml repo

MauriceHendrix committed:
adjust bad models

MauriceHendrix committed:
symlinks don't seem to work

MauriceHendrix committed:
abandon idea of symlinking

MauriceHendrix committed:
fix TestCvodeCells to new version of model

MauriceHendrix committed:
Update TestGenerateSteadyStateCrypt.hpp

Further relax number of cells check, due to variations in compiler / use of CVODE and differences in dependancies.

MauriceHendrix committed:
move to cmake_fetch for cellml files

MauriceHendrix committed:
fix TestModifiers to new location of cellml files

MauriceHendrix committed:
fix TestCvodeCells by finding cellml file in correct place

MauriceHendrix committed:
fix TestCvodeCells by chaging cellml file name

MauriceHendrix committed:
adjust TestCvodeCellsWithDataClamp to point to correct location for cellml files

MauriceHendrix committed:
remove extra space in cellml file path

MauriceHendrix committed:
Merge pull request #71 from Chaste/cellml_cmake_fetch

Cellml cmake fetch

MauriceHendrix commented:
@mirams I've merged the changes to use the cellml with cmake_fetch into here.

mirams commented on heart/test/ionicmodels/TestCvodeCells.hpp:
```diff
@@ -344,14 +350,14 @@ class TestCvodeCells : public CxxTest::TestSuite
         // Need to load dynamicly as we don't have config files anymore
            OutputFileHandler handler("TestCvodeCells", true);
 
```
Don't know whether these continuous tests that refer to the couple of CellML files in chaste source should be changing? There used to be just a handful of CellML files in the source so they got auto-converted on building (rather than all of them) and then those were used (and re-used in many test files) for all the continuous test pack.

mirams commented on heart/test/ionicmodels/TestCodegen.hpp:
```diff
@@ -159,8 +164,13 @@ class TestCodegen : public CxxTest::TestSuite
+        double sampling_interval = 1.0; // for testing codegen error check
 
         // Normal model
+        // test error from codegen about concentration
```
Might want to make these TS_ASSERT_THROWS_CONTAINS("Concentration JSR_calcium_concentration below 0") as they probably won't always agree to 6 decimal places.

MauriceHendrix committed:
use TS-ASSET_INCLUDES

mirams commented on crypt/test/simulation/TestGenerateSteadyStateCrypt.hpp:
```diff
@@ -169,7 +169,7 @@ class TestGenerateSteadyStateCrypt : public CxxTest::TestSuite
             // Define some "sensible" bounds on the number
             unsigned expected_cell_count_ub = 480u;
-            unsigned expected_cell_count_lb = 425u;
+            unsigned expected_cell_count_lb = 420u;
```
This shouldn't be changing in this commit? (Might need to merge in latest develop?)

mirams commented on heart/test/ionicmodels/TestCodegen.hpp:
```diff
@@ -315,6 +331,10 @@ class TestCodegen : public CxxTest::TestSuite
 #ifdef CHASTE_CVODE
         // CVODE version
+        // test error from codegen about concentration
+        Cellnegative_concentration_paci_hyttinen_aaltosetala_severi_ventricularVersionFromCellMLCvodeOpt cvode_model(p_solver, p_stimulus);
```
and all these ones too!

MauriceHendrix committed:
use TS-ASSET_INCLUDES

MauriceHendrix committed:
catching up with develop

MauriceHendrix committed:
revert to using model in src/odes for TestCvodeCells
